Migration to Kotlin

LilianaFaustinoBloco opened the pull request:
Full migration to kotlin.

Filmaluco commented on README.md:
```diff
@@ -3,7 +3,7 @@
 Generates fake data for testing or populating a development database.
```
com.github.blocoio:faker:1.2.9, I know you will update this after, but I guess this goes to 2.0.0...? or 1.3? Just a question :D

Filmaluco replied:
and the increase readme with the get started and examples. let me know if you want help

Filmaluco commented on lib/build.gradle.kts:
```diff
@@ -0,0 +1,32 @@
+        create<MavenPublication>("maven") {
+            groupId = "io.bloco"
+            artifactId = "faker"
+            version = "2.0.0"
```
ignore the one above :D answer is here :D

Filmaluco commented on lib/src/main/kotlin/io/bloco/faker/Faker.kt:
```diff
@@ -0,0 +1,111 @@
+                return input
+            }
+        } catch (e: IOException) {
+            //TODO
```
maybe a System.err.println("${e.stackTrace}")

Filmaluco commented on lib/src/main/kotlin/io/bloco/faker/components/Business.kt:
```diff
@@ -0,0 +1,24 @@
+        return fetch("business.credit_card_numbers")
+    }
+
+    fun creditCardExpireDate(): LocalDate {
```
@sdsantos just a question... this means it cant be use in KMM right? 
I know it would be a big change but...

Filmaluco replied:
like this there are other still using java

Filmaluco commented on lib/src/main/kotlin/io/bloco/faker/components/Company.kt:
```diff
@@ -0,0 +1,44 @@
+        val buzzwordsSections = getList("company", "buzzwords")
+        var catchPhrase = ""
+        for (buzzwordsSection in buzzwordsSections) {
+            catchPhrase += sampleFromList(buzzwordsSection as List<Any>) + " "
```
I am guessing this is List<String>?

Filmaluco commented on lib/src/main/kotlin/io/bloco/faker/components/Internet.kt:
```diff
@@ -0,0 +1,156 @@
+
+import io.bloco.faker.FakerComponent
+import io.bloco.faker.FakerData
+import java.util.*
```
avoid doing this imports :) I think theres a configuration in the IDE to avoid this

Filmaluco commented on lib/src/main/kotlin/io/bloco/faker/components/Internet.kt:
```diff
@@ -0,0 +1,156 @@
+
+class Internet(data: FakerData) : FakerComponent(data) {
+    fun email(name: String? = null): String {
+        return userName(name) + "@" + domainName()
```
I think you know but can use stringbuilder or "${method()} @ ${method()}"

Filmaluco commented on lib/src/main/kotlin/io/bloco/faker/helpers/RandomHelper.kt:
```diff
@@ -0,0 +1,76 @@
+        return random.nextInt(10).toString()
+    }
+
+    fun randBoolean(): Boolean {
```
I think most of this class can be used as extensions and or use the kotlin random library:
like https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.random/-random/next-boolean.html

Filmaluco commented on lib/src/test/kotlin/io/bloco/faker/components/AvatarTest.kt:
```diff
@@ -0,0 +1,47 @@
+import junit.framework.TestCase.assertFalse
+import org.junit.Before
+import org.junit.Test
+import kotlin.test.assertTrue
```
sometimes you use junit.framework.TestCase.assertTrue other kotlin.test.assertTrue :) Yes I am picking on small things... better sleep with my 👁️ 👁️ opened tonight...

Filmaluco commented on lib/src/test/kotlin/io/bloco/faker/components/InternetTest.kt:
```diff
@@ -0,0 +1,124 @@
+    @Test
+    fun userName() {
+        assertTrue(faker.internet.userName().matches(Regex("[\\w\\_\\.]+")))
+        assertTrue(faker.internet.userName("Sérgio Santos").matches(Regex("sergio[\\_\\.]santos")))
```
@sdsantos

Filmaluco commented on lib/src/test/kotlin/io/bloco/faker/components/LoremTest.kt:
```diff
@@ -0,0 +1,127 @@
+
+    @Before
+    fun setUp() {
+        faker = Faker()
```
I think you can use 
private val faker: Faker
   get() = Faker() 
 to always get a fresh one :) if you do inline val then for sure :D

Filmaluco commented on lib/src/main/kotlin/io/bloco/faker/helpers/StringHelper.kt:
```diff
@@ -0,0 +1,58 @@
+        fun replaceWith(matcher: Matcher): String
+    }
+
+    fun replaceMethod(input: String, regex: String, replacer: StringReplacer): String {
```
maybe change this to make use of invokes and extensions... better code. 
avoids the excess verbose on the tests

sdsantos commented on lib/src/test/kotlin/io/bloco/faker/FakerComponentTest.kt:
```diff
@@ -0,0 +1,94 @@
+    @Test
+    fun testFetchComposed() {
+        val componentKey = "wtv"
+        val options: List<String> = mutableListOf("John", "Mary")
```
Does not need to be mutable

sdsantos commented on lib/src/test/kotlin/io/bloco/faker/FakerComponentTest.kt:
```diff
@@ -0,0 +1,94 @@
+    fun testFetchComposed() {
+        val componentKey = "wtv"
+        val options: List<String> = mutableListOf("John", "Mary")
+        val internal: MutableMap<String, Any> = HashMap()
```
Can be `val internal = mutableMapOf<String, Any>()`

sdsantos commented on lib/src/test/kotlin/io/bloco/faker/components/AddressTest.kt:
```diff
@@ -0,0 +1,89 @@
+import org.junit.Test
+
+class AddressTest {
+    private lateinit var faker: Faker
+    @Before
+    fun setUp() {
+        faker = Faker()
+    }
```
It's fine to just do `val faker = Faker()` at the start of the class.

sdsantos commented on lib/src/test/kotlin/io/bloco/faker/components/BookTest.kt:
```diff
@@ -0,0 +1,34 @@
+
+    @Test
+    fun title() {
+        Assert.assertNotNull(faker.book.title())
```
We can import these statically.

sdsantos reviewed (commented):
Quick code review, mainly kotlin improvements.

LilianaFaustinoBloco commented:
@sdsantos 

I addressed all the comments and also tried to improve some parts of the code.
I ask your special attention to two things:

- Class StringHelper: I'm not sure if I achieved what was intended;
- Class Number, method `fun number(digits: Int): String` despite the tests never failing, I have doubts if it is the same as before.

Note: There are several unchecked casts throughout the code that I suggest we address in another PR.
